### DISCUSSION TITLE
Play to GStreamer or Icecast2

### DISCUSSION THREAD

**Discussion by h3ndrik** [music-assistant/support#2972] [Feature requests and ideas]
I would like to be able to stream the music to an Icecast2 internet radio server, which in turn forwards it to some hardware that can only play internet radio streams.

You already mention Icecast in the documentation in the section "I want to stream to Music Assistant" but that requires additional hardware and software. I would like some component that directly streams to an Icecast2 server.

A versatile way to implement this would be to leverage GStreamer. Which would be already integrated into Home Assistant via the respective [integration](https://www.home-assistant.io/integrations/gstreamer/). But I can't use that, since that integration [doesn't work on Home Assistant Core](https://github.com/home-assistant/core/issues/72875) and that hasn't been fixed in years.

We could define arbitrary gstreamer pipelines here (in Music Assisstant) and have it do various things, stream to Icecast, Snapcast, directly via TCP or whatever. I think that would be a good addition.

**Comment by OzGav** [comment]
An alternate solution has been identified here https://music-assistant.io/faq/stream-to/#web-radio

**Reply by rsmanning** [reply]
The problem with this solution is the ~8sec that isn't adjustable.  Its fine if you just want to pull something in but if you're trying to sync anything you have zero configuration options.  Would love to be able to locally source something in near real-time. Raw UDP reciever or ICE music source would be so nice.  Trying to figure out a way to get an Barix Instreamer 500 to work in realtime brought in as something other than a 8sec delay radio station.

**Comment by h3ndrik** [comment]
I can't get this to work. First, there is a step missing in the instructions. You also have to add the MPD to Home Assistant via the device integrations. Only after that it's ready to be found within Music Assistant. I don't really understand the "httpd_output" option. And unless I install this as a HA add-on, everything breaks. I've installed a Mopidy for debugging, and I can see some internal IPs from Home Assistant being passed over the MPD protocol. But that also doesn't work since these IPs and ports aren't reachable from outside.

I'd very much prefer a more general solution, like how we were used to set up Snapcast or Mopidy. Just a way to write to some FIFO or GStreamer pipeline, and we'll be able to connect it with anything. No matter if it's the local sound card, an internet stream or some network protocol via FFmpeg.

This alternate solution is overly complicated, and at least for me it doesn't seem to work.

**Comment by Hedda** [comment]
@h3ndrik in terminology specific to Music Assistent, is your original request asking developers to build a new "music provider", "player provider", "receiver plugin", or an "addon"?

* https://developers.music-assistant.io/

From your current description so far I can not even figure out if from where you want to stream and where you want to stream to.

Could you please explain in layman's terms what you would like to achieve (i.e. the goal, not only how to get there), and then maybe what could potentially be the architecture component needed, all explained in terminology specific to Music Assistent.

Is there perhaps also an existing "music provider", "player provider", "receiver plugin", or an "addon" for Music Assistent that does something similar which could be used as a template?

For reference, I am specifically wondering since OzGav mentioned this request in his pull request to add a webradio provider and from the original request here I can not tell if that is what you wanted as a solution for your needs:

* https://github.com/music-assistant/server/pull/4050
* 
PS: Should add that I do not believe that Music Assistent developers want it to be used as a pirate-radio tool, so I do think that any solution that make it super easy out-of-the-box to stream to everyone on the internet like an internet radio station server would be accepted into the main Music Assistent server codebase, (like just adding an Icecast Server would). As far as I know the Open Home Foundation goals for Music Assistent is still for it to only be a private tool for personal use and not a tool used for public broadcasting over the internet so anyone who finds out the URL can stream from it without authorization or authentication.
